File: app/src/main/java/wangdaye/com/geometricweather/main/utils/MainModuleUtils.java
package wangdaye.com.geometricweather.main.utils;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.view.View;

import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.utils.DisplayUtils;
import wangdaye.com.geometricweather.settings.SettingsOptionManager;

public class MainModuleUtils {

    private static final long BASE_ENTER_DURATION = 500;

    public static boolean needUpdate(Context context, Location location) {
        float pollingIntervalInHour = SettingsOptionManager.getInstance(context)
                .getUpdateInterval()
                .getIntervalInHour();
        return !location.isUsable()
                || location.getWeather() == null
                || !location.getWeather().isValid(pollingIntervalInHour);
    }

    public static Animator getEnterAnimator(View view, int pendingCount) {
        Animator[] animators = DisplayUtils.getFloatingOvershotEnterAnimators(view, 0.4f + 0.2f * pendingCount,
                DisplayUtils.dpToPx(view.getContext(), 120), 1.025f, 1.025f);

        AnimatorSet set = new AnimatorSet();
        set.playTogether(
                ObjectAnimator.ofFloat(view, "alpha", 0f, 1f),
                animators[0],
                animators[1],
                animators[2]
        );
        set.setDuration(Math.max(BASE_ENTER_DURATION - pendingCount * 50, BASE_ENTER_DURATION / 2));
        set.setStartDelay(pendingCount * 200);
        return set;
    }
}

File: app/src/main/java/org/breezyweather/main/utils/MainModuleUtils.kt
package org.breezyweather.main.utils

import android.animation.Animator
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.view.View
import org.breezyweather.common.utils.DisplayUtils
import kotlin.math.max

object MainModuleUtils {

    private const val BASE_ENTER_DURATION: Long = 500

    @JvmStatic
    fun getEnterAnimator(view: View, pendingCount: Int): Animator {
        val animators = DisplayUtils.getFloatingOvershotEnterAnimators(
            view, 0.4f + 0.2f * pendingCount,
            DisplayUtils.dpToPx(view.context, 120f), 1.025f, 1.025f
        )
        return AnimatorSet().apply {
            playTogether(
                ObjectAnimator.ofFloat(view, "alpha", 0f, 1f),
                animators[0],
                animators[1],
                animators[2]
            )
            duration = max(BASE_ENTER_DURATION - pendingCount * 50L, BASE_ENTER_DURATION / 2)
            startDelay = pendingCount * 200L
        }
    }
}
